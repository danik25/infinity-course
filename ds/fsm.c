/******************************************************************************
 * File name  : fsm.h
 * Version    : 1.1
 * Developer  : Dani Benjamin
 * Reviewer   : Yuval Shapira
 * Date		  : 28.4.19
 ******************************************************************************/

#include <stddef.h> /* size_t */
#include <assert.h> /* assert */
#include <stdio.h>

#include "fsm.h"

#define WORD_SIZE 8
#define PADDING(x) ((x >= WORD_SIZE) ? (WORD_SIZE - x % WORD_SIZE) : (WORD_SIZE - x))

struct fsm_manager
{
	size_t next_free;
	size_t block_size;
	size_t num_of_blocks;
};

/* block header struct */ 
typedef struct block_metadata
{	
	size_t offset;
}meta_t;

/***************************************************************************/

fsm_t *FSMInit(void *pool, size_t pool_size, size_t block_size)
{
	fsm_t *pool_handle = NULL;
	meta_t *runner = NULL;
	size_t increment = 0;
	size_t increment_track = 0;

	assert(pool);

	/*if the size isnt enough for the managment strucnt and one blloch header*/
	if(pool_size < sizeof(fsm_t) + sizeof(meta_t))
	{
		return (NULL);
	}

	/*assigning*/
	pool_handle = pool;
	runner = (meta_t*)pool + sizeof(fsm_t) / sizeof(meta_t);

	increment = block_size + PADDING(block_size) + sizeof(meta_t);
	increment_track = increment + sizeof(fsm_t);

	for(; (char*)runner < (char*)pool + pool_size - increment; 
										runner += increment / sizeof(meta_t))
	{
		runner->offset = increment_track; 
		increment_track += increment;
	}
	runner->offset = 0;

	/*pool handle*/
	pool_handle->next_free = *(char*)pool + sizeof(fsm_t);
	pool_handle->block_size = block_size;
	pool_handle->num_of_blocks = (pool_size - sizeof(fsm_t)) / increment;

	return (pool_handle);
}

/***************************************************************************/

void *FSMAlloc(fsm_t *fsm_handle)
{
	meta_t *runner = NULL;
	assert(fsm_handle);

	/*if there is no free blocks*/
	if((fsm_handle->next_free) == 0)
	{
		return(NULL);
	}

	/*swapping the data in the meta data strcut*/
	runner = (meta_t *)fsm_handle + fsm_handle->next_free / sizeof(meta_t);
	fsm_handle->next_free = runner->offset;
	runner->offset = (runner - (meta_t *)fsm_handle) * sizeof(meta_t);

	return ((char*)runner + sizeof(meta_t));
}

/***************************************************************************/

void FSMFree(void *block_addr)
{
	fsm_t *pool_handle = NULL;
	meta_t* runner = NULL;
	size_t tmp = 0;

	assert(block_addr);

	runner = (meta_t*)block_addr - sizeof(meta_t) / sizeof(meta_t);
	pool_handle = (fsm_t*)runner - runner->offset / sizeof(fsm_t);

	tmp = runner->offset;
	runner->offset = pool_handle->next_free;
	pool_handle->next_free = tmp;
}

/***************************************************************************/

size_t FSMCountFree(fsm_t *fsm_handle)
{
	meta_t* runner = NULL;
	size_t cnt = 1;

	assert(fsm_handle);

	runner = (meta_t*)fsm_handle + fsm_handle->next_free / sizeof(meta_t);
	while(runner->offset != 0)
	{
		++cnt;
		runner += fsm_handle->block_size + PADDING(fsm_handle->block_size)
										+sizeof(meta_t);
	}
	return (cnt);
}

/***************************************************************************/

size_t FSMSuggestSize(size_t block_num ,size_t block_size)
{
	return (block_num * (sizeof(meta_t) + block_size + PADDING(block_size)) + 
													sizeof(fsm_t));
}