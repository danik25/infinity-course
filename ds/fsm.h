/******************************************************************************
 * File name  : fsm.h
 * Version    : 1.1
 * Developer  : Dani Benjamin
 * Reviewer   : Yuval Shapira
 * Date		  : 28.4.19
 ******************************************************************************/
#ifndef __FSM_H__
#define __FSM_H__

#include <stddef.h> /* size_t */

/* API Description
 * FSM - Fixed Size Malloc, is a product that manages allocations of fixed sized
 * blocks. user provides the pool of allocated memory.
 */


typedef struct fsm_manager fsm_t;


/* description:
 *				initialize the memory pool provided by the user
 *              to fixed size blocks. 
 *
 * parameters: 
 *              pool - pointer to the user's allocated memory.
 *				pool_size - size of the user's allocated memory (in bytes).
 *				block_size - size of each block.
 *
 * return value: 
 *				pointer to the memory manager handler or
 *				NULL if failes.
 */				
fsm_t *FSMInit(void *pool, size_t pool_size, size_t block_size);


/* description:
 *				allocates a single block (with the size of block size). 
 *
 * parameters: 
 *              fsm_handle - pointer to the memory manager handler.
 *
 * return value: 
 *				pointer to the beginning of the allocated block
 *              NULL if no blocks available.
 */	
void *FSMAlloc(fsm_t *fsm_handle);


/* description:
 * 				frees a single block (with the size of block size). 
 *
 * parameters: 
 *              block_addr - pointer to the block to be freed.
 *
 * return value: 
 *				none.
 */	
void FSMFree(void *block_addr);


/* description:
 *				counts the number of free blocks in the pool.
 *
 * parameters: 
 *              fsm_handle - pointer to the memory manager handler.
 *
 * return value: 
 *				the number of free blocks.
 *
 *  complexity:
 *				O(n).
 */	
size_t FSMCountFree(fsm_t *fsm_handle);


/* description:
 *				calculates the suggested size of memory needed
 *			    to suit the user's demands. 
 *
 * parameters: 
 *              block_num - number of blocks in the pool.
 *				block_size - size of every block in the pool.
 *
 * return value: 
 *				suggested memory size to allocate (in bytes).
 */	
size_t FSMSuggestSize(size_t block_num ,size_t block_size);


#endif /* __FSM_H__ */ 