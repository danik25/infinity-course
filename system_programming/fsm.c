/******************************************************************************
 * File name  : fsm.h
 * Version    : 1.2
 * Developer  : Dani Benjamin
 * Reviewer   : Ayala Rosel
 * Date		  : 30.4.19
 ******************************************************************************/

#include <stddef.h> /* size_t */
#include <assert.h> /* assert */

#include "fsm.h"

#define END 0
#define WORD_SIZE 8
#define PADDING(x) (WORD_SIZE - x % WORD_SIZE) % WORD_SIZE 
                                                   

struct fsm_manager
{
	size_t next_free;
	size_t block_size;
	size_t num_of_blocks;
};

/* block header struct */ 
typedef struct block_metadata
{	
	size_t offset;
}meta_t;

/***************************************************************************/

fsm_t *FSMInit(void *pool, size_t pool_size, size_t block_size)
{
	fsm_t *pool_handle = NULL;
	meta_t *runner = NULL;
	size_t real_block_size = 0;
	size_t real_block_size_track = 0;
	size_t i = 0;

	assert(pool);

	/*if the size isnt enough for the managment strucnt and one blloch header*/
	if(pool_size < sizeof(fsm_t) + sizeof(meta_t) + block_size)
	{
		return (NULL);
	}
	/*assigning*/
	pool_handle = pool;
	runner = (meta_t*)((char*)pool + sizeof(fsm_t));

	real_block_size = block_size + PADDING(block_size) + sizeof(meta_t);
	real_block_size_track = real_block_size + sizeof(fsm_t);

	/*pool handle*/
	pool_handle->next_free = sizeof(fsm_t);
	pool_handle->block_size = block_size;
	pool_handle->num_of_blocks = (pool_size - sizeof(fsm_t)) / real_block_size;

	if((pool_size - real_block_size * pool_handle->num_of_blocks - 
							sizeof(fsm_t)) >= sizeof(meta_t) + block_size)
	{
		++pool_handle->num_of_blocks;
	}

	for(i = 1; i < pool_handle->num_of_blocks; ++i,
								runner += real_block_size / sizeof(meta_t))
	{
		runner->offset = real_block_size_track; 
		real_block_size_track += real_block_size;
	}
	runner->offset = END;

	return (pool_handle);
}

/***************************************************************************/

void *FSMAlloc(fsm_t *fsm_handle)
{
	meta_t *runner = NULL;

	assert(fsm_handle);

	/*if there is no free blocks*/
	if(END == (fsm_handle->next_free))
	{
		return(NULL);
	}

	/*swapping the data in the meta data strcut*/
	runner = (meta_t *)((char*)fsm_handle + fsm_handle->next_free);
	fsm_handle->next_free = runner->offset;
	runner->offset = (runner - (meta_t *)fsm_handle) * sizeof(meta_t);

	return ((char*)runner + sizeof(meta_t));
}

/***************************************************************************/

void FSMFree(void *block_addr)
{
	fsm_t *pool_handle = NULL;
	meta_t* runner = NULL;
	size_t tmp = 0;

	assert(block_addr);

	runner = (meta_t*)((char*)block_addr - sizeof(meta_t));
	pool_handle = (fsm_t*)((char*)runner - runner->offset);

	tmp = runner->offset;
	runner->offset = pool_handle->next_free;
	pool_handle->next_free = tmp;
}

/***************************************************************************/

size_t FSMCountFree(const fsm_t *fsm_handle)
{
	meta_t* runner = NULL;
	size_t cnt = 1;

	assert(fsm_handle);

	runner = (meta_t*)((char*)fsm_handle + fsm_handle->next_free);

    if(END == fsm_handle->next_free)
    {
        return(0);
    }

	while(runner->offset != END)
	{
		++cnt;
		runner = (meta_t*)((char*)fsm_handle +runner->offset);
	}
	return (cnt);
}


/***************************************************************************/

size_t FSMSuggestSize(size_t block_num ,size_t block_size)
{
	return (block_num * (sizeof(meta_t) + block_size + PADDING(block_size)) + 
													sizeof(fsm_t));
}