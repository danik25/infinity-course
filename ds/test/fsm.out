/home/xyz/git/dani-benjamin/ds/fsm.out